src/server/mod.rs

```diff
@@ -198,68 +258,102 @@ where T: AsyncRead + AsyncWrite,
+                        return Ok(PollMain::Done.into());
+                    },
+                    Err(err) => {
+                        trace!("service closed");
+                        // service is closed, transition to goaway state
+                        break Error::Service(err);
+                    }
+                }
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        if !self.is_ready() {
-            // Advance the initialization of the service and HTTP/2.0 connection
-            try_ready!(self.try_ready());
-        }
+                let next = connection.poll()
+                    .map_err(Error::Protocol);
 
-        let (connection, service) = match self.state {
-            State::Ready { ref mut connection, ref mut service } => {
-                (connection, service)
-            }
-            _ => unreachable!(),
-        };
+                let (request, respond) = match try_ready!(next) {
+                    Some(next) => next,
+                    None => return Ok(PollMain::Done.into()),
+                };
 
-        loop {
-            // Make sure the service is ready
-            let ready = service.poll_ready()
-                // TODO: Don't dump the error
-                .map_err(Error::Service);
+                let (parts, body) = request.into_parts();
 
-            try_ready!(ready);
+                // This is really unfortunate, but the `http` currently lacks the
+                // APIs to do this better :(
+                let mut request = Request::from_parts(parts, ());
+                self.modify.modify(&mut request);
 
-            let next = connection.poll()
-                .map_err(Error::Protocol);
+                let (parts, _) = request.into_parts();
+                let request = Request::from_parts(parts, RecvBody::new(body));
 
-            let (request, respond) = match try_ready!(next) {
-                Some(next) => next,
-                None => return Ok(().into()),
-            };
+                // Dispatch the request to the service
+                let response = service.call(request);
 
-            let (parts, body) = request.into_parts();
+                // Spawn a new task to process the response future
+                if let Err(_) = self.executor.execute(Background::new(respond, response)) {
+                    break Error::Execute;
+                }
+            }
+            _ => unreachable!(),
+        };
 
-            // This is really unfortunate, but the `http` currently lacks the
-            // APIs to do this better :(
-            let mut request = Request::from_parts(parts, ());
-            self.modify.modify(&mut request);
+        // We only break out of the loop on a service error, which means we
+        // should transition to GOAWAY.
+        match mem::replace(&mut self.state, State::Done) {
+            State::Ready { mut connection, .. } => {
+                connection.abrupt_shutdown(Reason::INTERNAL_ERROR);
```

Comment:
Why do you opt for `abrupt_shutdown` here?

Reply:
This state means either: a) there was an error returned from `Service::poll_ready`, or b) there was an error from `Executor::execute`. I kind of assumed that meant something really bad was happening. Should I not?

Reply:
Hmm, so if `Service::poll_ready` returning an error does not mean that already open connections cannot complete, only that the service cannot accept **new** requests.

`Executor::execute` can fail when when it is performing a graceful shutdown. In this case, the executor does not accept **new** tasks, but can continue to process already running ones.

Also, less critically for now, `Executor::execute` can fail w/ a transient error (out of capacity). In this case, we'd probably want to implement some back off strategy.

I would probably use a graceful shutdown in this case.